src/wallet/coinselection.cpp

```diff
@@ -48,20 +49,33 @@ bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_va
         remaining += utxo.txout.nValue;
     }
 
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<std::pair<bool, bool>> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
     // Depth first search to find
     while (!done)
     {
         if (tries <= 0) { // Too many tries, exit
-            return false;
+            break;
         } else if (value_ret > target_value + cost_of_change) { // Selected value is out of range, go back and try other branch
             backtrack = true;
+        } else if (curr_waste > best_waste) { // Don't select things which we know will be more wasteful
+            backtrack = true;
         } else if (value_ret >= target_value) { // Selected value is within range
-            done = true;
+            curr_waste += value_ret - target_value;
```

Comment:
parens around the value being subtracted would be nice

Reply:
Done here and elsewhere